ccbuilder/compilerstore.py

```diff
@@ -0,0 +1,178 @@
+from dataclasses import dataclass
+from pathlib import Path
+import json
+
+from ccbuilder.utils.utils import CompilerProject
+from ccbuilder.utils.repository import Repo
+
+
+@dataclass(frozen=True, slots=True)
+class BuiltCompilerInfo:
+    project: CompilerProject
+    prefix: Path
+    revision: str
+
+    # def get_compiler(self, cpp:bool=False)-> CompilerExe:...
+
+
+@dataclass(kw_only=True)
+class CompilerStore:
+    built_gcc_compilers: dict[str, BuiltCompilerInfo]
+    built_llvm_compilers: dict[str, BuiltCompilerInfo]
+
+    # def get_built_compilers(self, project: CompilerProject) -> list[BuiltCompiler]:
+    # match project:
+    # case CompilerProject.GCC:
+    # return list(self.built_gcc_compilers.values())
+    # case CompilerProject.LLVM:
+    # return list(self.built_llvm_compilers.values())
+
+    def add_compiler(self, compiler: BuiltCompilerInfo) -> None:
+        if compiler.project == CompilerProject.GCC:
+            self.built_gcc_compilers[compiler.revision] = compiler
+        elif compiler.project == CompilerProject.LLVM:
+            self.built_llvm_compilers[compiler.revision] = compiler
+
+    def get_built_compiler(
+        self, project: CompilerProject, revision: str
+    ) -> BuiltCompilerInfo | None:
+        match project:
+            case CompilerProject.GCC:
+                return self.built_gcc_compilers.get(revision, None)
+            case CompilerProject.LLVM:
+                return self.built_llvm_compilers.get(revision, None)
```

Comment:
I think there could be a check if the compiler really does exist and if not update the db accordingly.
A check for if the compiler really does *not* exist seems more important.

The first check might be a bit expensive, but should not be too bad as the directory will be accessed afterwards anyway.
The second check seems cheap and will also keep the db in sync. Having a discrepancy in the DB can already happen with multiprocessing.

Reply:
Supporting multiprocessing in general is a headache

Reply:
Like
```
result = self.built_gcc_compilers.get(revision, None)

if res is None and store.changed_since_last_scanned():
    // Check if the compiler I think does not exist exists and update accordingly
```
?

Reply:
Yup, on the other directly checking for the compiler in the filesystem might make more sense

Reply:
So I switch it to using an actual DB, this should be a non-issue. 

Expect in the following case: 
- Someone is almost done building a compiler
- Another process queries whether that compiler is already built the result is False
- Right after that the first process commits the built
- Now the second process tries to built the same compiler again

I don't know how realistic this is but I think it should be caught at another level: the second process when attemting to build the compiler realized someone else has already built it/is currently building it

Reply:
So I added the worker_pid check back in. A special exception is thrown `DuplicateBuildException`, so the client can busy way if it wants to.